### PR TITLE
fix(draw): use byte as stride unit instead of pixel

### PR DESCRIPTION
**Pull request by XuNeo**
### Description of the feature or fix

For image source that stride is aligned to certain bytes, use pixel as draw unit will fail.

### Checkpoints
- [x] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [ ] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [ ] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [ ] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [ ] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [ ] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [ ] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [ ] Use typed pointers instead of `void *` pointers
- [ ] Do not `malloc` into a static or global variables. Instead declare the variable in `lv_global_t` structure in [`lv_global.h`](https://github.com/lvgl/lvgl/blob/master/src/core/lv_global.h) and mark the variable with `(LV_GLOBAL_DEFAULT()->variable)` when it's used. See a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#memory-management).
- [ ] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [ ] Widget members function must start with `lv_<module_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [ ] `struct`s should be used via an API and not modified directly via their elements.
- [ ] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [ ] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [ ] Arguments must be named in H files too.
- [ ] To register and use callbacks one of the following needs to be followed (see a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#callbacks)):
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

### CODE REVIEW

**Comment by kisvegabor** [comment]
I was working on the same here: 2a5f9ecb2

Does it solve your issue too?

**Comment by kisvegabor** [comment]
For the sake of simplicity I divide the stride by 2 (for RGB565) or 4  (for XRGB8888) but we can do that as the stride should be always 2^N aligned.

**Comment by XuNeo** [comment]
> Does it solve your issue too?

No, it's failing on image with manually aligned stride. 
I have rebased the code. The principle is that for both src, and dest, treat them all as draw buf and using stride to goto next row.

**Comment by XuNeo** [comment]
Emmm, there's still conflict. Do you want me to hold on until you finish it?

**Comment by kisvegabor** [comment]
Finished :slightly_smiling_face: 
 
> No, it's failing on image with manually aligned stride.

Can you mention an image width, and stride combination which doesn't work? Or event better, could you send an image that I can test width?

**Comment by XuNeo** [comment]
I'm testing with `cogwheel.RGB888.bin`.
It can be generated with commands below. I'm testing with stride aligned to 16bytes. Indexed image won't work for now. I'm checking it.

`./scripts/LVGLImage.py -v --ofmt BIN --cf AUTO  --align 16 ../pngs`
[pngs.zip](https://github.com/lvgl/lvgl/files/13364391/pngs.zip)

**Comment by kisvegabor** [comment]
For RGB888 I used the stride in byte during blending. 

I think the best would be to add all image format tests to `LV_DEMO_RENDER_IMAGE_NORMAL`. Only 4 images can fit on a page so we can create `LV_DEMO_RENDER_IMAGE_1_NORMAL`, `LV_DEMO_RENDER_IMAGE_2_NORMAL`, etc.

I can do this on the weekend but if you have time to deal with it earlier it's would be great. Of course we should use the stride=64 images as it was discussed in #4624. 

FYI, I [added render tests](https://github.com/lvgl/lvgl/tree/master/tests/ref_imgs/draw/render) for RGB565, RGB888, XRGB888, and ARGB8888 targets. So by adding more image format tests to `lv_demo_render` we should have a pretty good coverage.

**Comment by XuNeo** [comment]
Agree to add more render test. I'm currently looking the issue of stride process in bin image decoder, and will prepare another PR when finished.

Once it's done, then another PR for more renderer test. 
I still have other promises on our meeting like adding compression support to bin etc. Please do add the test if you have time.

**Comment by kisvegabor** [comment]
I started to work on the image format tests but realized that, indexed, alpha only, RLE, etc are related to image decoding and not purely image rendering. This way they are the same for all GPUs and renderers.

With this in mind keeping them separated from the rendering tests makes more sense to me.  I tink the current tests with the cogwheel are fine, however probably we should replace the cogwheel with something else which has more colors and less symmetrical. We could use a smaller image too, in order to have more tests on a screen.

While working on it I've noticed that you treat the stride of RGB565A8 differently than I imagined. It seems you use 3 bytes pixel size and calculate the stride from it. However using 3 bytes we cannot navigate in this image format as we have 2 bytes/pixels in the RGB565 part and 1 byte/pixel in he A8 part. 

It's a little bit hacky but what I did in SW render is setting the stride for the RGB565 part, and assume half of it for the A8 part.

**Comment by XuNeo** [comment]
> With this in mind keeping them separated from the rendering tests makes more sense to me. I tink the current tests with the cogwheel are fine, however probably we should replace the cogwheel with something else which has more colors and less symmetrical. We could use a smaller image too, in order to have more tests on a screen.

Agree. Need also add different width such as 127x127 that none-aligned.

How would you provide different configuration combinations in CI test? For example, `LV_BIN_DECODER_RAM_LOAD` or not.


> While working on it I've noticed that you treat the stride of RGB565A8 differently than I imagined. It seems you use 3 bytes pixel size and calculate the stride from it. However using 3 bytes we cannot navigate in this image format as we have 2 bytes/pixels in the RGB565 part and 1 byte/pixel in he A8 part.

I'm not sure about it. I tend to specify stride separately. Because once we introduce YUV format, it will meet same issue. See below struct, where`uv_stride`, `v_stride`, `alpha_stride` could be different.

Instead of converting stride unit from byte to pixel (2 bytes/pixels for map, 1 byte/pixel for alpha), it may be a good idea to say alpha stride is equal to map stride in bytes for now. The python converter script can take care of it. In the end when separate stride parameters are supported, then it's a easy migration.

```
    typedef struct vg_lite_yuvinfo {
        vg_lite_swizzle_t   swizzle;        /* UV swizzle. */
        vg_lite_yuv2rgb_t   yuv2rgb;        /* 601 or 709 conversion standard. */
        uint32_t            uv_planar;      /* UV(U) planar address. */
        uint32_t            v_planar;       /* V planar address. */
        uint32_t            alpha_planar;   /* Alpha planar address. */
        uint32_t            uv_stride;      /* UV(U) stride. */
        uint32_t            v_stride;       /* V stride. */
        uint32_t            alpha_stride;   /* Alpha stride. */
        uint32_t            uv_height;      /* UV(U) height. */
        uint32_t            v_height;       /* V height. */
        void *              uv_memory;      /* The logical pointer to the UV(U) planar memory. */
        void *              v_memory;       /* The logical pointer to the V planar memory. */
        void *              uv_handle;      /* The memory handle of the UV(U) planar. */
        void *              v_handle;       /* The memory handle of the V planar. */
    } vg_lite_yuvinfo_t;
```

**Comment by kisvegabor** [comment]
> Agree. Need also add different width such as 127x127 that none-aligned.

I vote for something smaller and more arbitrary, like, 37x41

> How would you provide different configuration combinations in CI test? For example, LV_BIN_DECODER_RAM_LOAD or not.

Now we are running the unit tests with 2 configurations. The main difference is that one uses stdlib malloc, the other uses tlsf `malloc`, however I also wee that we need more config options for:
- `LV_BIN_DECODER_RAM_LOAD`
- Parallel rendering
- Float support
- Anything else?

We need to check which one of these needs to have their own tests and which can be combined with the others. I thinking about this:
1. no parallel, no os, no float, TLSF: test TLSF based malloc, stdlib malloc is superior (because it support ASAN) so we should use it is the other tests
2. no parallel, pthread, float, stdlib malloc: 
3. 8 threads, pthread,  some other configs, stdlib malloc
4. ? threads, pthread,  some other configs, stdlib malloc

> Instead of converting stride unit from byte to pixel (2 bytes/pixels for map, 1 byte/pixel for alpha), it may be a good idea to say alpha stride is equal to map stride in bytes for now. The python converter script can take care of it. In the end when separate stride parameters are supported, then it's a easy migration.

So e.g. for a 30x20 image, with stride align = 32 the stride for RGB would be 30x2 = 60 -> 64, and we would use the  64 for the A8 part as well, although 32 would be enough?  If so I think it's too wasteful even for a temporary solution. It looks like having 4 bytes per pixel instead of 3, so 33% increase. On a 200x200 image, it's 120kB vs 160kB.

**Comment by XuNeo** [comment]
> Now we are running the unit tests with 2 configurations. The main difference is that one uses stdlib malloc, the other uses tlsf `malloc`, however I also wee that we need more config options for:

I agree with your config combination. We also need to include different stride alignment. Probably `LV_DRAW_BUF_STRIDE_ALIGN` and `LV_DRAW_BUF_ALIGN`.

1. base: no parallel, no os, no float, stdlib malloc, LV_DRAW_BUF_STRIDE_ALIGN = 0
2. base + LV_DRAW_BUF_STRIDE_ALIGN = 64 + float + TLSF
3. base + LV_DRAW_BUF_STRIDE_ALIGN = 64 + float + `LV_BIN_DECODER_RAM_LOAD=1`( need ASAN to verify 2)
4. base + 8 threads


> I vote for something smaller and more arbitrary, like, 37x41
Do you have an candidate to replace current `cogwheel`?



> So e.g. for a 30x20 image, with stride align = 32 the stride for RGB would be 30x2 = 60 -> 64, and we would use the 64 for the A8 part as well, although 32 would be enough? If so I think it's too wasteful even for a temporary solution. It looks like having 4 bytes per pixel instead of 3, so 33% increase. On a 200x200 image, it's 120kB vs 160kB.

RGB565A8 seems hard to follow a consistent definition for stride and bit-per-pixel. Let's see if we can have better options like eliminating this format by converting it to RGB565 + alpha map mask after decoding. So draw unit can process it in a more standard color format, unless GPU specifically wants RGB565A8 via decoder open argument.

As for now, I can firstly make it work with the calculated alpha stride.


For this PR, I tested ARGB8888 binary image with master branch and it seems working now. Is this PR still good to merge?

**Comment by kisvegabor** [comment]
> I agree with your config combination. We also need to include different stride alignment. Probably LV_DRAW_BUF_STRIDE_ALIGN and LV_DRAW_BUF_ALIGN.

Sounds good. Let's discuss it in a separate issue.

> Do you have a candidate test image to replace current cogwheel? Please update it and then new test case can use it.

I will search something tomorrow. 


> RGB565A8 seems hard to follow a consistent definition for stride and bit-per-pixel.

I think we can define our own rules for it, like a custom image format. I believe using RGB stride / 2 for A8 stride is safe for GPUs as well.

> For this PR, I tested ARGB8888 binary image with master branch and it seems working now. Is this PR still good to merge? The idea is to unify all usage of stride and make sure unit is byte to avoid converting from pixel to byte.

Your solution is really more unified and less hacky, so I merge it. :blush:
